traversal/planner/PlannerEdge.java

```diff
@@ -129,47 +140,59 @@ public String toString() {
             extends TraversalEdge<VERTEX_DIR_FROM, VERTEX_DIR_TO> {
 
         OptimiserVariable.Boolean varIsSelected;
-        OptimiserVariable.Boolean[] varOrderAssignment;
-        private OptimiserVariable.Integer varOrderNumber;
+        OptimiserVariable.Boolean varIsMinimal;
+        OptimiserConstraint conIsMinimal;
+
         private final String varPrefix;
         private final String conPrefix;
         private final GraphPlanner planner;
         private final Encoding.Direction.Edge direction;
-        private boolean hasInitialValue;
-        private boolean isInitialisedVariables;
-        private boolean isInitialisedConstraints;
-        private double costNext;
-        double costLastRecorded;
+
+        private boolean didInitialiseVariables;
+        private boolean didInitialiseConstraints;
+        private boolean hasInitialValues;
+
+        private final int tieBreaker;
+        private static int nextTieBreaker = 0;
```

Comment:
Needed so that there can be only one minimal incoming edge into a vertex at a time.

Reply:
let's say the DB is around for a loooong time and doesn't get restarted - Java doesn't do overflows per say but it's still awkward to permit them. I'd say we should still promote this to a `long`!
Also this currently isn't thread safe: two DirectionalEdges that are created in parallel can have the same tie breaker (we don't currently do this within the same query, but it's possible). Because it's static we should use an `AtomicLong`, or choose a different strategy...
choosing a random number still has a chance of collision huh? Same with a hash...

The other option is that we can create the `PlannerEdge` with an ID (long) we generate when creating the `GraphPlanner`, which it can pass into the `Directional` edges as a disambiguator.
This is probably actually the cleanest solution, what do you think?

Reply:
We only care that the tie breakers are distinct within a single planner's set of edges, and given that the planner is not threaded, there is no danger of collisions. That said, AtomicLong seems to be a cleaner solution overall, with only a tiny and extremely rare penalty during the planner setup.
I don't think involving the planner in determining the tie breakers—or ids, or whatever we call them—is a good idea, because that exposes an implementation detail that the planner really shouldn't know anything about.